### PR TITLE
hyperv: improve networking code

### PR DESCRIPTION
**Pull request by Maners**
* make VM XML network interface always use 
````xml
<interface type="bridge"><source bridge="GUID"/></interface>
````
for both define and dumpxml actions
* implement virConectListAllNetworks, virNetworkGetXMLDesc and other networking methods so that we can get lists of Hyper-V networks by getting network handles and then get network UUIDs (as names may be duplicated)

### CODE REVIEW

File: src/hyperv/hyperv_network_api_v1.c [hunk 1]
```diff
@@ -264,6 +264,81 @@ hyperv1ConnectNumOfNetworks(virConnectPtr conn)
...
+    char *filter = NULL;
+    char uuid_string[VIR_UUID_STRING_BUFLEN];
+    virNetworkDefPtr def;
+
+    if (VIR_ALLOC(def) < 0)
+        return NULL;
+
+    virUUIDFormat(network->uuid, uuid_string);
+
+    if (virAsprintf(&filter, "Name = \"%s\"", uuid_string) < 0 ||
```
Comment (SriRamanujam) [inline src/hyperv/hyperv_network_api_v1.c]: I think you missed the actual call to `hyperv1VirtualSwitch` here.

File: src/hyperv/hyperv_api_v1.c [hunk 2]
```diff
@@ -1663,13 +1671,11 @@ hyperv1DomainAttachSyntheticEthernetAdapter(virDomainPtr domain,
...
 
     /* build the __PATH variable of the switch port */
-    if (hyperv1MsvmComputerSystemFromDomain(domain, &computer) < 0)
-        goto cleanup;
     if (virAsprintf(&connection__PATH, "\\\\%s\\root\\virtualization:"
                 "Msvm_SwitchPort.CreationClassName=\"Msvm_SwitchPort\","
                 "Name=\"%s\",SystemCreationClassName=\"Msvm_VirtualSwitch\","
-                "SystemName=\"%s\"", computer->data->ElementName,
-                switchport_guid_string, net->data.network.name) < 0)
+                "SystemName=\"%s\"", hostname,
```
Comment (SriRamanujam) [inline src/hyperv/hyperv_api_v1.c:1669]: Is the hostname from Libvirt always guaranteed to be what Windows expects?

File: src/hyperv/hyperv_api_v1.c [hunk 3]
```diff
@@ -1594,20 +1594,28 @@ hyperv1DomainAttachSyntheticEthernetAdapter(virDomainPtr domain,
...
-    Msvm_ComputerSystem_V1 *computer = NULL;
+    Msvm_VirtualSwitch_V1 *vSwitch = NULL;
     eprParam virtualSwitch_REF, ComputerSystem_REF;
     simpleParam virtualSwitch_Name, virtualSwitch_FriendlyName,
                 virtualSwitch_ScopeOfResidence;
     embeddedParam ResourceSettingData;
     invokeXmlParam *params = NULL;
     properties_t *NewResources = NULL;
 
+    /* get the vSwitch first so we can get data from it as needed */
```
Comment (SriRamanujam) [inline src/hyperv/hyperv_api_v1.c]: Why is this necessary? The only thing you need from this is the GUID of the Msvm_VirtualSwitch object, which is already present in the XML (`net->data.bridge.brname`). You even use it in the error message on line 1658.

File: src/hyperv/hyperv_api_v1.c [hunk 4]
```diff
@@ -1594,20 +1594,28 @@ hyperv1DomainAttachSyntheticEthernetAdapter(virDomainPtr domain,
...
+    if (hyperv1GetMsvmVirtualSwitchList(priv, &query, &vSwitch) < 0 || vSwitch == NULL)
+        goto cleanup;
+
     /*
      * step 1: create virtual switch port
      * https://msdn.microsoft.com/en-us/library/cc136782(v=vs.85).aspx
      */
     virBufferAddLit(&query, MSVM_VIRTUALSWITCH_V1_WQL_SELECT);
-    virBufferAsprintf(&query, "where Name = \"%s\"", net->data.network.name);
+    virBufferAsprintf(&query, "where Name = \"%s\"", vSwitch->data->Name);
```
Comment (SriRamanujam) [inline src/hyperv/hyperv_api_v1.c]: Should just use the GUID from the XML.

File: src/hyperv/hyperv_api_v1.c [hunk 5]
```diff
@@ -1663,13 +1671,11 @@ hyperv1DomainAttachSyntheticEthernetAdapter(virDomainPtr domain,
...
     /* build the __PATH variable of the switch port */
-    if (hyperv1MsvmComputerSystemFromDomain(domain, &computer) < 0)
-        goto cleanup;
     if (virAsprintf(&connection__PATH, "\\\\%s\\root\\virtualization:"
                 "Msvm_SwitchPort.CreationClassName=\"Msvm_SwitchPort\","
                 "Name=\"%s\",SystemCreationClassName=\"Msvm_VirtualSwitch\","
-                "SystemName=\"%s\"", computer->data->ElementName,
-                switchport_guid_string, net->data.network.name) < 0)
+                "SystemName=\"%s\"", hostname,
+                switchport_guid_string, vSwitch->data->Name) < 0)
```
Comment (SriRamanujam) [inline src/hyperv/hyperv_api_v1.c]: Should just use the GUID from the XML.

File: src/hyperv/hyperv_api_v2.c [hunk 6]
```diff
@@ -1716,22 +1715,19 @@ hyperv2DomainAttachSyntheticEthernetAdapter(virDomainPtr domain,
...
 
-    if (hyperv2MsvmComputerSystemFromDomain(domain, &computer) < 0)
-        goto cleanup;
-
     /* build the two __PATH variables */
     if (virAsprintf(&switch__PATH, "\\\\%s\\root\\virtualization\\v2:"
                 "Msvm_VirtualEthernetSwitch.CreationClassName="
                 "\"Msvm_VirtualEthernetSwitch\",Name=\"%s\"",
-                computer->data->ElementName, net->data.network.name) < 0)
+                hostname, vSwitch->data->Name) < 0)
```
Comment (SriRamanujam) [inline src/hyperv/hyperv_api_v2.c:1722]: Is the hostname always guaranteed to be what Windows expects?

File: src/hyperv/hyperv_network_api_v1.c [hunk 7]
```diff
@@ -71,104 +71,142 @@ hyperv1MsvmVirtualSwitchToNetwork(virConnectPtr conn,
     return result;
 }
 
-/* exported API functions */
-
-int
-hyperv1ConnectListNetworks(virConnectPtr conn, char **const names, int maxnames)
+static int
+hyperv1GetVirtualSwitchList(hypervPrivate *priv,
+                            const char *filter,
```
Comment (SriRamanujam) [inline src/hyperv/hyperv_network_api_v1.c:76]: You don't clean up this filter string anywhere. It should probably be cleaned up in this function.

File: src/hyperv/hyperv_network_api_v1.c [hunk 8]
```diff
@@ -71,104 +71,142 @@ hyperv1MsvmVirtualSwitchToNetwork(virConnectPtr conn,
...
         goto cleanup;
+    }
 
-    for (virtualSwitch = virtualSwitchList; virtualSwitch != NULL;
-         virtualSwitch = virtualSwitch->next) {
-        count++;
+    if (hyperv1GetVirtualSwitchList(priv, NULL, &vSwitchList) < 0)
+        goto cleanup;
+
+    if (networks) {
```
Comment (SriRamanujam) [inline src/hyperv/hyperv_network_api_v1.c]: This entire block seems oddly laid out. I would first allocate `nets` big enough to hold all the networks (you have this number from `hyperv1GetVirtualSwitchList`), then run through the loop below without the call to `VIR_RESIZE_N`.

File: src/hyperv/hyperv_network_api_v1.c [hunk 9]
```diff
@@ -71,104 +71,142 @@ hyperv1MsvmVirtualSwitchToNetwork(virConnectPtr conn,
...
 
-    result = count;
+    for (vSwitch = vSwitchList; vSwitch != NULL; vSwitch = vSwitch->next) {
+        /* just count the networks */
+        if (!networks) {
+            count++;
+            continue;
+        }
+
+        if (VIR_RESIZE_N(nets, nnets, count, 2) < 0)
```
Comment (SriRamanujam) [inline src/hyperv/hyperv_network_api_v1.c]: Allocate the networks all at once above, don't resize in here.

File: src/hyperv/hyperv_network_api_v1.c [hunk 10]
```diff
@@ -71,104 +71,142 @@ hyperv1MsvmVirtualSwitchToNetwork(virConnectPtr conn,
...
-    int result = -1, count = 0;
     hypervPrivate *priv = conn->privateData;
-    virBuffer query = VIR_BUFFER_INITIALIZER;
-    Msvm_VirtualSwitch_V1 *virtualSwitchList = NULL;
-    Msvm_VirtualSwitch_V1 *virtualSwitch = NULL;
+    Msvm_VirtualSwitch_V1 *vSwitchList = NULL;
+    Msvm_VirtualSwitch_V1 *vSwitch = NULL;
+    virNetworkPtr network = NULL;
+    virNetworkPtr *nets = NULL;
+    int count = 0;
```
Comment (SriRamanujam) [inline src/hyperv/hyperv_network_api_v1.c:117]: Minor nitpick: this should be `size_t`.

File: src/hyperv/hyperv_network_api_v2.c [hunk 11]
```diff
@@ -71,103 +71,141 @@ hyperv2MsvmVirtualSwitchToNetwork(virConnectPtr conn,
     return result;
 }
 
-/* exported API functions */
-
-int
-hyperv2ConnectListNetworks(virConnectPtr conn, char **const names, int maxnames)
+static int
+hyperv2GetVirtualSwitchList(hypervPrivate *priv,
+                            const char *filter,
```
Comment (SriRamanujam) [inline src/hyperv/hyperv_network_api_v2.c:76]: Free this variable in the cleanup.

File: src/hyperv/hyperv_network_api_v2.c [hunk 12]
```diff
@@ -71,103 +71,141 @@ hyperv2MsvmVirtualSwitchToNetwork(virConnectPtr conn,
...
-    int result = -1, count = 0;
     hypervPrivate *priv = conn->privateData;
-    virBuffer query = VIR_BUFFER_INITIALIZER;
-    Msvm_VirtualEthernetSwitch_V2 *virtualSwitchList = NULL;
-    Msvm_VirtualEthernetSwitch_V2 *virtualSwitch = NULL;
+    Msvm_VirtualEthernetSwitch_V2 *vSwitchList = NULL;
+    Msvm_VirtualEthernetSwitch_V2 *vSwitch = NULL;
+    virNetworkPtr network = NULL;
+    virNetworkPtr *nets = NULL;
+    int count = 0;
```
Comment (SriRamanujam) [inline src/hyperv/hyperv_network_api_v2.c:117]: Minor nitpick: this should be `size_t`.

File: src/hyperv/hyperv_network_api_v2.c [hunk 13]
```diff
@@ -71,103 +71,141 @@ hyperv2MsvmVirtualSwitchToNetwork(virConnectPtr conn,
...
         goto cleanup;
+    }
 
-    for (virtualSwitch = virtualSwitchList; virtualSwitch != NULL;
-         virtualSwitch = virtualSwitch->next) {
-        count++;
+    if (hyperv2GetVirtualSwitchList(priv, NULL, &vSwitchList) < 0)
+        goto cleanup;
+
+    if (networks) {
```
Comment (SriRamanujam) [inline src/hyperv/hyperv_network_api_v2.c]: Why not just allocate the full list of networks in this block and skip the `VIR_RESIZE_N` call in the loop below?

File: src/hyperv/hyperv_network_api_v2.c [hunk 14]
```diff
@@ -71,103 +71,141 @@ hyperv2MsvmVirtualSwitchToNetwork(virConnectPtr conn,
...
+        if (VIR_ALLOC_N(nets, 1) < 0)
+            goto cleanup;
+        nnets = 1;
     }
 
-    result = count;
+    for (vSwitch = vSwitchList; vSwitch != NULL; vSwitch = vSwitch->next) {
+        /* just count the networks */
+        if (!networks) {
```
Comment (SriRamanujam) [inline src/hyperv/hyperv_network_api_v2.c]: This block should probably be outside of the loop, since you know how many networks there are via `hyperv2GetVirtualSwitchList`.

File: src/hyperv/hyperv_network_api_v1.c [hunk 15]
```diff
@@ -71,104 +71,142 @@ hyperv1MsvmVirtualSwitchToNetwork(virConnectPtr conn,
...
+    if (networks) {
+        if (VIR_ALLOC_N(nets, 1) < 0)
+            goto cleanup;
+        nnets = 1;
     }
 
-    result = count;
+    for (vSwitch = vSwitchList; vSwitch != NULL; vSwitch = vSwitch->next) {
+        /* just count the networks */
+        if (!networks) {
```
Comment (SriRamanujam) [inline src/hyperv/hyperv_network_api_v1.c]: This block should probably outside of this loop, since you know how many networks there are via `hyperv1GetVirtualSwitchList`.

File: src/hyperv/hyperv_network_api_v2.c [hunk 16]
```diff
@@ -71,103 +71,141 @@ hyperv2MsvmVirtualSwitchToNetwork(virConnectPtr conn,
...
+hyperv2ConnectListNetworks(virConnectPtr conn, char **const names, int maxnames)
 {
-    int i, count = 0;
-    bool success = false;
+    Msvm_VirtualEthernetSwitch_V2 *list = NULL, *entry = NULL;
     hypervPrivate *priv = conn->privateData;
-    virBuffer query = VIR_BUFFER_INITIALIZER;
-    Msvm_VirtualEthernetSwitch_V2 *virtualSwitchList = NULL;
-    Msvm_VirtualEthernetSwitch_V2 *virtualSwitch = NULL;
+    int count = 0;
```
Comment (SriRamanujam) [inline src/hyperv/hyperv_network_api_v2.c:189]: Minor nitpick: this should be `size_t`.

File: src/hyperv/hyperv_network_api_v2.c [hunk 17]
```diff
@@ -191,57 +228,53 @@ hyperv2NetworkLookupByName(virConnectPtr conn, const char *name)
...
+        return NULL;
 
-    virBufferAddLit(&query, MSVM_VIRTUALETHERNETSWITCH_V2_WQL_SELECT);
-    virBufferAsprintf(&query, "where Description = \"%s\" and ElementName = \"%s\"",
-                      "Microsoft Virtual Switch", name);
-    if (hyperv2GetMsvmVirtualEthernetSwitchList(priv, &query,
-                &virtualSwitch) < 0) {
-        goto cleanup;
-    }
     if (virtualSwitch == NULL) {
```
Comment (SriRamanujam) [inline src/hyperv/hyperv_network_api_v2.c]: Combine this with the block above: 
```C
if (hyperv2GetVirtualSwitchList(priv, filter, &virtualSwitch) || virtualSwitch == NULL) {
    virReportError(VIR_ERR_NO_NETWORK, _("No network found with name %s"), name);
    return NULL;
}
```

File: src/hyperv/hyperv_network_api_v1.c [hunk 18]
```diff
@@ -264,6 +264,81 @@ hyperv1ConnectNumOfNetworks(virConnectPtr conn)
...
+    hypervFreeObject(priv, (hypervObject *) vSwitch);
+    virNetworkDefFree(def);
+
+    return xml;
+}
+
+int
+hyperv1NetworkSetAutostart(virNetworkPtr network ATTRIBUTE_UNUSED,
+                           int autostart)
+{
```
Comment (SriRamanujam) [inline src/hyperv/hyperv_network_api_v1.c:306]: Since this function is a no-op, why not always return -1?

File: src/hyperv/hyperv_network_api_v1.h [hunk 19]
```diff
@@ -32,6 +32,11 @@ int hyperv1ConnectListDefinedNetworks(virConnectPtr conn, char **const names,
         int maxnames);
 virNetworkPtr hyperv1NetworkLookupByName(virConnectPtr conn, const char *name);
 int hyperv1ConnectNumOfDefinedNetworks(virConnectPtr conn);
+char * hyperv1NetworkGetXMLDesc(virNetworkPtr network, unsigned int flags);
```
Comment (SriRamanujam) [inline src/hyperv/hyperv_network_api_v1.h]: Should be `char *hyperv1NetworkGetXMLDesc`

**Review by SriRamanujam — CHANGES_REQUESTED** [review]
I don't mean to be too nitpicky, but yeah. As usual, let me know if I missed anything or didn't pick up on some nuance.

File: src/hyperv/hyperv_api_v1.c (hunk 3 above)
Comment (Maners) [reply src/hyperv/hyperv_api_v1.c]: yeah that's left over from the previous iteration when I had the code work with network name instead of guid - this was looking up the vSwitch by name and then using guild down the line.

File: src/hyperv/hyperv_network_api_v1.c (hunk 7 above)
Comment (Maners) [reply src/hyperv/hyperv_network_api_v1.c:76]: this string is stack allocated in all callers, so there's nothing to clean-up

File: src/hyperv/hyperv_network_api_v1.c (hunk 18 above)
Comment (Maners) [reply src/hyperv/hyperv_network_api_v1.c:306]: you can enable autostart but not disable - like in the esx driver.

File: src/hyperv/hyperv_network_api_v2.c (hunk 11 above)
Comment (Maners) [reply src/hyperv/hyperv_network_api_v2.c:76]: again, this is stack-allocated by the caller so once it gets out of scope it will cleanup

File: src/hyperv/hyperv_network_api_v2.c (hunk 12 above)
Comment (Maners) [reply src/hyperv/hyperv_network_api_v2.c:117]: return type of this function is int and it eventually gets assigned to ret

File: src/hyperv/hyperv_network_api_v2.c (hunk 16 above)
Comment (Maners) [reply src/hyperv/hyperv_network_api_v2.c:189]: same as above, its used as return value on success so it should match the function's return type which is int

File: src/hyperv/hyperv_api_v1.c (hunk 2 above)
Comment (Maners) [reply src/hyperv/hyperv_api_v1.c:1669]: the hostname is passed from hypervNDomainDefineXML and is result of hypervNGetHostSystem call which in turn is Msvm_ComputerSystem->data->ElementName so we had 2 lookups here instead of using what we already have given to us by the caller.

File: src/hyperv/hyperv_api_v2.c (hunk 6 above)
Comment (Maners) [reply src/hyperv/hyperv_api_v2.c:1722]: yes,  it's pulled from Msvm_Computersystem->data->ElementName by the caller and passed to this one as hostname

File: src/hyperv/hyperv_network_api_v1.c (hunk 1 above)
Comment (Maners) [reply src/hyperv/hyperv_network_api_v1.c]: yep, fixed.

File: src/hyperv/hyperv_network_api_v1.c (hunk 8 above)
Comment (Maners) [reply src/hyperv/hyperv_network_api_v1.c]: fixed.

**Comment by Maners** [comment]
@SriRamanujam thanks for reviewing :-) I've addressed all the issues and force-pushed to keep clean commit history, except:

* the variable passed as const char *filter does not need cleanup as it's not heap allocated
* hostname is safe to use as it's fetch from Msvm_ComputerSystem by the calling code already so there's no need to lookup same info twice
* size_t for count variable should not be used because count is used as return value (assigned to "ret" right before cleanup label) of functions that declare int return type.

**Review by SriRamanujam — APPROVED** [review]
